### PR TITLE
@W-21148602: Internal Server List does not respect changes made to Servers.xml file

### PR DESCRIPTION
**Pull request by JohnsonEricAtSalesforce**
🎸 _*Ready For Review*_ 🥁

  This updates the login server manager to respond to updates made to the login servers from servers.xml in resources and the login servers provided by Mobile Device Management via the runtime configuration.  Previously, login servers from these two locations were static (for servers.xml) or additions-only (for runtime config) after the app had been installed.

  For a point release, these changes do need to maintain compatibility with the 13.x public API.  To minimize regression risk, they're also maintaining the Java style and deprecated shared preferences persistence already in place.

  Accomplishing the change is non-trivial.  The current logic was not built with mutability in mind for the non-custom servers.  Scaling the existing logic and ensuring the shared preferences data maintains exact integrity did increase the change footprint.

  New unit tests provide functional automated testing of all the changes and 100% patch code coverage.  The overall test coverage for the login server manager is much, much higher than before as well.

  I'll follow-up with inline source commentary as well.  Thanks for reading!

### CODE REVIEW

File: libs/SalesforceSDK/src/com/salesforce/androidsdk/config/LoginServerManager.java [hunk 1]
```diff
@@ -92,7 +91,14 @@ public LoginServerManager(Context ctx) {
 				Context.MODE_PRIVATE);
 		runtimePrefs = ctx.getSharedPreferences(RUNTIME_PREFS_FILE,
 				Context.MODE_PRIVATE);
+
+		// Reset non-custom servers from mobile device management (MDM) and servers XML.
+		resetNonCustomLoginServers(runtimePrefs);
```
Comment (JohnsonEricAtSalesforce) [inline libs/SalesforceSDK/src/com/salesforce/androidsdk/config/LoginServerManager.java]: @wmathurin, @brandonpage and @bbirman, before I update the tests and move this into ready-for-review I wanted to run the concept of the change past you all.

I chatted briefly with @bbirman to review the iOS behavior since the work item specifically hails that logic as working compared to Android's.  Currently, MSDK Android loads the `servers.xml` only once on install.  The work item mentions upgrade, but I didn't see that logic.  More, when MDM provides servers they're additive only.  When MDM drops a server, the app doesn't reflect that.

What I'm trying here and seems to manually test well is:
1. Drop both sets of non-custom servers on startup, including non-custom servers from `servers.xml` and MDM.
2. Load the current set of non-custom servers from either `servers.xml` _or_ MDM according to the existing logic, as much as possible.
3. When MDM is not enabled, preserve the user's custom login servers at the end of the list

I'm doing a little more testing around the behavior of the only show authorized hosts parameter.  I want to be sure the MDM flow still respects the user's custom servers.

The question to answer is: Does this seem like a reasonable update to the existing logic to resolve this work item? I can manipulate `servers.xml` and the MDM data and the app seems to keep the servers list adds, updates or removes non-custom servers just as we'd want.  The logic here is older and the methods are not as self-contained as our newer logic, so I wanted to be sure we all got to review any potential side-effects of this before moving further!

Comment (bbirman) [reply libs/SalesforceSDK/src/com/salesforce/androidsdk/config/LoginServerManager.java]: For #3 (and pending your authorized hosts testing), on iOS the custom login servers would be preserved if MDM was enabled as long as "onlyShowAuthorizedHosts" is false

Comment (brandonpage) [reply libs/SalesforceSDK/src/com/salesforce/androidsdk/config/LoginServerManager.java]: This sounds good to me.  

One thought though: Should we only remove MDM servers when network is available?  Security wise, it would be bad it:

1. Put device in airplane mode
2. Open app.  MDM servers are deleted and updated servers cannot be retrieved.  
3. Turn airplane mode off.
4. `LoginServerManager` constructor does not run again.  MDM server list is not respected.

Is that scenario possible?

Comment (JohnsonEricAtSalesforce) [reply libs/SalesforceSDK/src/com/salesforce/androidsdk/config/LoginServerManager.java]: Thanks and that thought about network is a great topic.  I'll do some research!

Comment (JohnsonEricAtSalesforce) [reply libs/SalesforceSDK/src/com/salesforce/androidsdk/config/LoginServerManager.java]: I took a manual look and also asked our agents.  RuntimeConfig does cache the policy locally and will give MSDK those values when no network is present ✅

File: libs/SalesforceSDK/src/com/salesforce/androidsdk/config/LoginServerManager.java [hunk 2]
```diff
@@ -68,31 +73,90 @@ public class LoginServerManager {
 	public static final String WELCOME_LOGIN_URL = "https://welcome.salesforce.com/discovery";
 	public static final String SANDBOX_LOGIN_URL = "https://test.salesforce.com";
 
-	// Keys used in shared preferences.
```
Comment (JohnsonEricAtSalesforce) [inline libs/SalesforceSDK/src/com/salesforce/androidsdk/config/LoginServerManager.java:71]: I've reviewed the diff in Github once already, so my pre-amble is that Github does a terrible job (as it often does) correlating the diff.  If one really wants to navigate the diff more easily, use Android Studio to compare the revisions 🎉

File: libs/SalesforceSDK/src/com/salesforce/androidsdk/config/LoginServerManager.java [hunk 3]
```diff
@@ -68,31 +73,90 @@ public class LoginServerManager {
...
-	private static final String SERVER_URL_FILE = "server_url_file";
-	private static final String RUNTIME_PREFS_FILE = "runtime_prefs_file";
-	private static final String NUMBER_OF_ENTRIES = "number_of_entries";
-	private static final String SERVER_NAME = "server_name_%d";
-	private static final String SERVER_URL = "server_url_%d";
-	private static final String IS_CUSTOM = "is_custom_%d";
-	private static final String SERVER_SELECTION_FILE = "server_selection_file";
+	/**
+	 * Shared preferences when non-custom login servers are provided by resources servers.xml
```
Comment (JohnsonEricAtSalesforce) [inline libs/SalesforceSDK/src/com/salesforce/androidsdk/config/LoginServerManager.java:77]: There are some new documentation comments to help many of the less-intuitively named properties and methods be clearer and easily reviewed in Android Studio.

File: libs/SalesforceSDK/src/com/salesforce/androidsdk/config/LoginServerManager.java [hunk 4]
```diff
@@ -68,31 +73,90 @@ public class LoginServerManager {
...
+	 */
+	@VisibleForTesting
```
Comment (JohnsonEricAtSalesforce) [inline libs/SalesforceSDK/src/com/salesforce/androidsdk/config/LoginServerManager.java:79]: `@VisibleForTesting` appears in many places to enable the new tests.

File: libs/SalesforceSDK/src/com/salesforce/androidsdk/config/LoginServerManager.java [hunk 5]
```diff
@@ -68,31 +73,90 @@ public class LoginServerManager {
...
+
+	/**
+	 * The Android context used for shared preferences access
+	 */
 	private final Context ctx;
+
+	/**
+	 * The Android runtime configuration used for Mobile Device Management (MDM)
+	 */
+	private final RuntimeConfig runtimeConfig;
```
Comment (JohnsonEricAtSalesforce) [inline libs/SalesforceSDK/src/com/salesforce/androidsdk/config/LoginServerManager.java:114]: Only two properties were added or modified: `runtimeConfig` is one and present for internal testing support.

File: libs/SalesforceSDK/src/com/salesforce/androidsdk/config/LoginServerManager.java [hunk 6]
```diff
@@ -68,31 +73,90 @@ public class LoginServerManager {
...
+
+	/** The resource id of the servers.xml file */
+	private final int serversXmlResourceId;
```
Comment (JohnsonEricAtSalesforce) [inline libs/SalesforceSDK/src/com/salesforce/androidsdk/config/LoginServerManager.java:117]: Only two properties were added or modified: `serversXmlResourceId` is one and present for internal testing support.

File: libs/SalesforceSDK/src/com/salesforce/androidsdk/config/LoginServerManager.java [hunk 7]
```diff
@@ -68,31 +73,90 @@ public class LoginServerManager {
...
+		this(ctx, getRuntimeConfig(ctx), servers);
+	}
+
+	/**
+	 * Constructs a new login server manager.
+	 *
+	 * @param ctx The context
+	 */
+	@VisibleForTesting
+	public LoginServerManager(
```
Comment (JohnsonEricAtSalesforce) [inline libs/SalesforceSDK/src/com/salesforce/androidsdk/config/LoginServerManager.java:144]: This new and private parameterized constructor is present for testing dependency injection.

File: libs/SalesforceSDK/src/com/salesforce/androidsdk/config/LoginServerManager.java [hunk 8]
```diff
@@ -68,31 +73,90 @@ public class LoginServerManager {
...
-				Context.MODE_PRIVATE);
-		runtimePrefs = ctx.getSharedPreferences(RUNTIME_PREFS_FILE,
-				Context.MODE_PRIVATE);
+		this.runtimeConfig = runtimeConfig;
+		this.serversXmlResourceId = serversXmlResourceId;
+		settings = ctx.getSharedPreferences(SERVER_URL_FILE, MODE_PRIVATE);
+		runtimePrefs = ctx.getSharedPreferences(RUNTIME_PREFS_FILE, MODE_PRIVATE);
+
+		// (Re-)initialize non-custom login servers provided by the resources servers.xml.
+		resetNonCustomLoginServers(settings);
```
Comment (JohnsonEricAtSalesforce) [inline libs/SalesforceSDK/src/com/salesforce/androidsdk/config/LoginServerManager.java:156]: A crux change is here in the constructor to (re-)initialize the non-custom login servers from resources `server.xml` when applicable.  Note, this doesn't affect the user's custom login servers.

File: libs/SalesforceSDK/src/com/salesforce/androidsdk/config/LoginServerManager.java [hunk 9]
```diff
@@ -68,31 +73,90 @@ public class LoginServerManager {
...
 		initSharedPrefFile();
+
+		// Select a default login server.
 		getSelectedLoginServer();
```
Comment (JohnsonEricAtSalesforce) [inline libs/SalesforceSDK/src/com/salesforce/androidsdk/config/LoginServerManager.java:160]: The updated `getSelectedLoginServer` has new logic to detect when the "selected" login server is no longer in the "available" login servers and select the new default login server.

File: libs/SalesforceSDK/src/com/salesforce/androidsdk/config/LoginServerManager.java [hunk 10]
```diff
@@ -118,39 +182,57 @@ public LoginServer getLoginServerFromURL(String url) {
...
-		} else {
+		// Fetch the selected login server.
+		final SharedPreferences selectedServerPrefs = ctx.getSharedPreferences(SERVER_SELECTION_FILE, MODE_PRIVATE);
+		final String selectedServerName = selectedServerPrefs.getString(SERVER_NAME, null);
+		final String selectedServerUrl = selectedServerPrefs.getString(SERVER_URL, null);
+		final boolean selectedServerIsCustom = selectedServerPrefs.getBoolean(IS_CUSTOM, false);
+		final boolean selectedServerHasValidNameAndUrl = selectedServerName != null && selectedServerUrl != null;
+
+		// Refresh the list of mobile device management (MDM) servers from the runtime config.
+		if (isRuntimeConfigAppServiceHostsSet()) {
```
Comment (JohnsonEricAtSalesforce) [inline libs/SalesforceSDK/src/com/salesforce/androidsdk/config/LoginServerManager.java:199]: This refresh of the runtime config login servers wasn't present before but now nicely compliments the following call to `getLoginServers` so the selected login server will reflect when MDM has removed the currently selected login server.  This was an odd looking bug where the "selected" login server might no longer appear in the server picker after an MDM update.

File: libs/SalesforceSDK/src/com/salesforce/androidsdk/config/LoginServerManager.java [hunk 11]
```diff
@@ -118,39 +182,57 @@ public LoginServer getLoginServerFromURL(String url) {
...
+			getLoginServersFromRuntimeConfig();
+		}
+
+		// Get the active list of login servers.
+		final List<LoginServer> loginServers = getLoginServers();
+
+		// Check if the selected login server is available in the active list of login servers.
+		final boolean selectedLoginServerIsAvailable = loginServers.stream().anyMatch(server ->
```
Comment (JohnsonEricAtSalesforce) [inline libs/SalesforceSDK/src/com/salesforce/androidsdk/config/LoginServerManager.java:207]: Our agent tooling recommended this `stream` approach to search the list in Java.  This'll be easier in Kotlin, someday.

File: libs/SalesforceSDK/src/com/salesforce/androidsdk/config/LoginServerManager.java [hunk 12]
```diff
@@ -187,12 +268,12 @@ public void useSandbox() {
...
-	 * Adds a custom login server to the shared pref file.
+	 * Adds a custom login server.
 	 *
-	 * @param name Server name.
-	 * @param url Server URL.
+	 * @param name The login server name
+	 * @param url The login server URL
 	 */
-	public void addCustomLoginServer(String name, String url) {
+	public void addCustomLoginServer(@NonNull String name, @NonNull String url) {
```
Comment (JohnsonEricAtSalesforce) [inline libs/SalesforceSDK/src/com/salesforce/androidsdk/config/LoginServerManager.java:276]: Though it hadn't been reported, passing null for these parameters was a potential `NullPointerException`.

File: libs/SalesforceSDK/src/com/salesforce/androidsdk/config/LoginServerManager.java [hunk 13]
```diff
@@ -201,11 +282,12 @@ public void addCustomLoginServer(String name, String url) {
 			}
 		}
 
-		if (getLoginServersFromRuntimeConfig() == null) {
-			persistLoginServer(name, url, true, settings);
-		} else {
-			persistLoginServer(name, url, true, runtimePrefs);
-		}
+		persistLoginServer(
```
Comment (JohnsonEricAtSalesforce) [inline libs/SalesforceSDK/src/com/salesforce/androidsdk/config/LoginServerManager.java:285]: The new `getSharedPreferences()` method lets this be one call to `persistLoginServer` with the correct `SharedPreferences`.

File: libs/SalesforceSDK/src/com/salesforce/androidsdk/config/LoginServerManager.java [hunk 14]
```diff
@@ -219,109 +301,152 @@ public void reset() {
...
-	 * @param server the server to remove
+	 * @param server                The server to remove
+	 * @param sharedPreferences     The shared preferences to remove the server from
+	 * @param allowNonCustomRemoval Boolean true allows the removal of non-custom login servers and
+	 *                              false does not
 	 */
-	public void removeServer(LoginServer server) {
-		List<LoginServer> servers = getLoginServers();
+	@VisibleForTesting
+	public void removeServer(
```
Comment (JohnsonEricAtSalesforce) [inline libs/SalesforceSDK/src/com/salesforce/androidsdk/config/LoginServerManager.java:329]: This private and parameterized version of `removeServer` supports testing.

File: libs/SalesforceSDK/src/com/salesforce/androidsdk/config/LoginServerManager.java [hunk 15]
```diff
@@ -219,109 +301,152 @@ public void reset() {
...
 	 */
 	public List<LoginServer> getLoginServers() {
-		List<LoginServer> allServers = getLoginServersFromRuntimeConfig();
-		if (allServers == null) {
-			allServers = getLoginServersFromPreferences();
-		} else {
-			allServers = getLoginServersFromPreferences(runtimePrefs);
-		}
-		return allServers;
+		return getLoginServersFromPreferences(getSharedPreferences());
```
Comment (JohnsonEricAtSalesforce) [inline libs/SalesforceSDK/src/com/salesforce/androidsdk/config/LoginServerManager.java:373]: `getLoginServers` also gets a lot simpler since it can have one call to `getLoginServersFromPreferences` with the correct `SharedPreferences` parameter.

File: libs/SalesforceSDK/src/com/salesforce/androidsdk/config/LoginServerManager.java [hunk 16]
```diff
@@ -219,109 +301,152 @@ public void reset() {
...
 				SalesforceSDKLogger.w(TAG, "No login servers labels provided or wrong number of login servers labels provided - using URLs for the labels");
 				mdmLoginServersLabels = mdmLoginServers;
 			}
-			final List<LoginServer> storedServers = getLoginServersFromPreferences(runtimePrefs);
-			for (int i = 0; i < mdmLoginServers.length; i++) {
-				final String name = mdmLoginServersLabels[i];
-				final String url = mdmLoginServers[i];
+
+			// Reset non-custom servers from Mobile Device Management (MDM).
+			resetNonCustomLoginServers(runtimePrefs);
```
Comment (JohnsonEricAtSalesforce) [inline libs/SalesforceSDK/src/com/salesforce/androidsdk/config/LoginServerManager.java:416]: Another crux change is here in `getLoginServersFromRuntimeConfig` to reset the non-custom login servers from that `SharedPreferences` so that strange things don't happen as the MDM data changes.  My testing from `dev` showed that adding servers worked but deleting servers had no effect.  This also meant that "editing" a server resulted in a duplication.  The list always reflects MDM now.

File: libs/SalesforceSDK/src/com/salesforce/androidsdk/config/LoginServerManager.java [hunk 17]
```diff
@@ -219,109 +301,152 @@ public void reset() {
...
+
+			// Null-cleanse MDM login server URLs and names.
+			final List<String> mdmLoginServersList = asList(mdmLoginServers);
+			mdmLoginServersList.removeIf(Objects::isNull);
```
Comment (JohnsonEricAtSalesforce) [inline libs/SalesforceSDK/src/com/salesforce/androidsdk/config/LoginServerManager.java:420]: Our agent tools recommended this approach for null-cleansing the MDM data which aided in simplifying the tests and avoiding code coverage noise.

File: libs/SalesforceSDK/src/com/salesforce/androidsdk/config/LoginServerManager.java [hunk 18]
```diff
@@ -349,28 +474,16 @@ public void reorderCustomLoginServer(
...
+		final List<LoginServer> servers = getLoginServers();
+		updatedIndex = getIndexAdjustedToCustomLoginServerBounds(updatedIndex, getSharedPreferences());
 
 		// Update the login server list.
 		loginServers.remove(originalIndex);
 		loginServers.add(updatedIndex, originalLoginServer);
 
 		// Edit each login server indexed after the updated index.
-		final Editor editor = settings.edit();
+		final Editor editor = getSharedPreferences().edit();
```
Comment (JohnsonEricAtSalesforce) [inline libs/SalesforceSDK/src/com/salesforce/androidsdk/config/LoginServerManager.java:486]: This was a latent bug where the wrong `SharedPreferences` was hard-coded even when MDM was active.

File: libs/SalesforceSDK/src/com/salesforce/androidsdk/config/LoginServerManager.java [hunk 19]
```diff
@@ -414,6 +527,32 @@ public void replaceCustomLoginServer(
...
 	}
 
+	/**
+	 * Adjusts a login server index to be within the bounds of the custom login servers.
+	 *
+	 * @param index The login server index
+	 * @param sharedPreferences The login server shared preferences
+	 * @return The adjusted login server index
+	 */
+	private @NonNull Integer getIndexAdjustedToCustomLoginServerBounds(
```
Comment (JohnsonEricAtSalesforce) [inline libs/SalesforceSDK/src/com/salesforce/androidsdk/config/LoginServerManager.java:537]: These utility methods to adjust a login server's index to the appropriate bounds of non-custom or custom servers are useful, but for future when `SharedPreferences` is replaced it'd be amazing to see the custom servers stored separately from the non-custom ones entirely.

File: libs/SalesforceSDK/src/com/salesforce/androidsdk/config/LoginServerManager.java [hunk 20]
```diff
@@ -435,83 +574,142 @@ private List<LoginServer> getLegacyLoginServers() {
...
-					eventType = xml.next();
-				} catch (XmlPullParserException | IOException e) {
-					SalesforceSDKLogger.w(TAG, "Exception thrown while parsing XML", e);
-				}
+			}
+			try {
+				eventType = xml.next();
+			} catch (Exception e) {
+				SalesforceSDKLogger.w(TAG, "Exception thrown while parsing XML", e);
+				break;
```
Comment (JohnsonEricAtSalesforce) [inline libs/SalesforceSDK/src/com/salesforce/androidsdk/config/LoginServerManager.java:602]: Without this break, this exception turned into infinite recursion and an app hang which I caught in testing while adding coverage to this (previously?) untested line.

File: libs/SalesforceSDK/src/com/salesforce/androidsdk/config/LoginServerManager.java [hunk 21]
```diff
@@ -435,83 +574,142 @@ private List<LoginServer> getLegacyLoginServers() {
...
 		}
 		int numServers = servers.size();
 		final Editor edit = settings.edit();
 		for (int i = 0; i < numServers; i++) {
 			final LoginServer curServer = servers.get(i);
-			edit.putString(String.format(Locale.US, SERVER_NAME, i), curServer.name.trim());
-			edit.putString(String.format(Locale.US, SERVER_URL, i), curServer.url.trim());
-			edit.putBoolean(String.format(Locale.US, IS_CUSTOM, i), curServer.isCustom);
-			if (i == 0) {
+			persistLoginServer(
```
Comment (JohnsonEricAtSalesforce) [inline libs/SalesforceSDK/src/com/salesforce/androidsdk/config/LoginServerManager.java:642]: All login server persistence happens now via `persistLoginServer` since it has all the logic needed to correctly order and sanitize the `SharedPreferences`.

File: libs/SalesforceSDK/src/com/salesforce/androidsdk/config/LoginServerManager.java [hunk 22]
```diff
@@ -435,83 +574,142 @@ private List<LoginServer> getLegacyLoginServers() {
...
+					curServer.isCustom,
+					settings
+			);
+
+			// Set the default login server to the first entry once.
+			if (i == 0 && ctx.getSharedPreferences(SERVER_SELECTION_FILE, MODE_PRIVATE).getAll().isEmpty()) {
 				setSelectedLoginServer(curServer);
 			}
 		}
-		edit.putInt(NUMBER_OF_ENTRIES, numServers);
```
Comment (JohnsonEricAtSalesforce) [inline libs/SalesforceSDK/src/com/salesforce/androidsdk/config/LoginServerManager.java:492]: This `NUMBER_OF_ENTRIES` update happens in `persistLoginServer` now.

File: libs/SalesforceSDK/src/com/salesforce/androidsdk/config/LoginServerManager.java [hunk 23]
```diff
@@ -435,83 +574,142 @@ private List<LoginServer> getLegacyLoginServers() {
...
+	 * @param name              The login server name
+	 * @param url               The login server URL
+	 * @param isCustom          boolean true for non-custom (managed) login servers, false for
+	 *                          custom (user-entered) login servers
+	 * @param sharedPreferences The shared preferences
 	 */
-	private void persistLoginServer(String name, String url, boolean isCustom, SharedPreferences sharedPrefs) {
-		if (name == null || url == null) {
-			return;
+	private void persistLoginServer(final @NonNull String name,
```
Comment (JohnsonEricAtSalesforce) [inline libs/SalesforceSDK/src/com/salesforce/androidsdk/config/LoginServerManager.java:666]: Another big crux change is here to `persistLoginServer` which now has the ability to correctly order non-custom and custom login servers and all the other login servers.  This is where the most work actually happened to make the non-custom login server list mutable.

File: libs/SalesforceSDK/src/com/salesforce/androidsdk/config/LoginServerManager.java [hunk 24]
```diff
@@ -520,22 +718,40 @@ private void persistLoginServer(String name, String url, boolean isCustom, Share
 	 * @param prefs SharedPreferences file.
 	 * @return List of all saved servers.
 	 */
-	private List<LoginServer> getLoginServersFromPreferences(SharedPreferences prefs) {
+	@VisibleForTesting
+	public @NonNull List<LoginServer> getLoginServersFromPreferences(final SharedPreferences prefs) {
 		int numServers = prefs.getInt(NUMBER_OF_ENTRIES, 0);
 		if (numServers == 0) {
-			return null;
+			return new ArrayList<>();
```
Comment (JohnsonEricAtSalesforce) [inline libs/SalesforceSDK/src/com/salesforce/androidsdk/config/LoginServerManager.java:725]: I left most other methods returning `null` if the did on `dev` but this one actually made sense to eliminate the null checks at all the private call sites and simplify the logic and tests.

File: libs/test/SalesforceSDKTest/src/com/salesforce/androidsdk/auth/LoginServerManagerTest.kt [hunk 25]
```diff
@@ -0,0 +1,817 @@
...
+import org.junit.Assert.assertEquals
+import org.junit.Assert.assertNull
+import org.junit.Before
+import org.junit.Test
+import org.junit.runner.RunWith
+import org.xmlpull.v1.XmlPullParserException
+
+@RunWith(AndroidJUnit4::class)
+@SmallTest
+class LoginServerManagerTestKt {
```
Comment (JohnsonEricAtSalesforce) [inline libs/test/SalesforceSDKTest/src/com/salesforce/androidsdk/auth/LoginServerManagerTest.kt]: Where `mockk` was needed, I wrote new tests directly in Kotlin.

**Comment by JohnsonEricAtSalesforce** [comment]
Testing Notes: I tested this via both manual testing on device and via all the new unit tests.  In particular, I also used a temporary method to log the contents of both `SharedPreferences` during testing so I could be 100% sure the data looked rock solid.  That was really tedious, but it did reveal interesting data quality issues that unit tests and even manual testing could miss - there was one bug where data indexed _above_ the number of servers value was orphaned after edits.  The unit tests and even the app didn't detect that, but I saw it by putting my own eyes on it ✅

Testing requires restarting the app after:
1. Adding to `servers.xml`
2. Editing `servers.xml`
3. Removing from `servers.xml`
...And since I didn't have a managed device for the MDM tests I had to update the values obtained from `RuntimeConfig` in the code to...
4. Add servers
5. Edit servers
6. Remove servers

Each of these tests should be tested with a cold start of the app to ensure the (re-)initialization of the `LoginServerManager` correctly restores the servers.

Each of these tests should be performed with and without the presence of custom login servers.

The unit tests provide more destructive testing, such as inserting null values and malformed XML in all kinds of places they shouldn't be.  I spent five days in manual testing and automated testing development since I'd view this as a higher risk change.  This was where I found and resolved a number of `NullPointerException` crashes and one infinite hang of the app when XML is malformed. 😅

File: libs/test/SalesforceSDKTest/src/com/salesforce/androidsdk/auth/LoginServerManagerTest.kt (hunk 25 above)
Comment (brandonpage) [reply libs/test/SalesforceSDKTest/src/com/salesforce/androidsdk/auth/LoginServerManagerTest.kt]: Maybe rename this one to `LoginServerManagerMockTest`?  

At the very least let's make the file name match the class so we don't have `LoginServerManageTest.java` and `LoginServerManageTest.kt`.

**Review by brandonpage — APPROVED** [review]

**Review by wmathurin — APPROVED** [review]
